### ISSUE TITLE
using npm module in packaged up jxcore binary

### ISSUE THREAD

**Issue by pirog** [jxcore/jxcore#542]
So we are working on a project over at https://github.com/kalabox/kalabox

Getting close to rolling out a useable release and have been considering easier ways to distribute. Thinking about the option of compiling into a binary lead us here. 

Using `jx` seems to work pretty well but there are 2 big issues that are roadblocking us. Both are documented at https://github.com/kalabox/kalabox/issues/464#issuecomment-139907519.

The first is easily solvable but the second seems much more elusive and strange hence the ticket here for some guidance. 

At the end of the day we need to programmatically copy the skeleton of a node project over into another directory and then `npm install`. The `npm install` part is the tricky part. We have npm packaged up inside the generated binary but it seems to crash with the following when it tries to do the programmatic `npm install`. 

```
ENOENT, no such file or directory '/usr/local/bin/node_modules/npm/node_modules/npm-registry-client/lib'
Error: ENOENT, no such file or directory '/usr/local/bin/node_modules/npm/node_modules/npm-registry-client/lib'
    at Object.fs.readdirSync (evalmachine.<anonymous>:717:20)
    at CachingRegistryClient.RegClient (/usr/local/bin/node_modules/npm/node_modules/npm-registry-client/index.js.jx:49:6)
    at new CachingRegistryClient (/usr/local/bin/node_modules/npm/lib/cache/caching-client.js.jx:19:18)
    at /usr/local/bin/node_modules/npm/lib/npm.js.jx:362:22
    at /usr/local/bin/node_modules/npm/lib/config/core.js.jx:81:7
    at Array.forEach (native)
    at /usr/local/bin/node_modules/npm/lib/config/core.js.jx:80:13
    at f (/usr/local/bin/node_modules/npm/node_modules/once/once.js.jx:16:25)
    at finalize (/usr/local/bin/node_modules/npm/lib/config/core.js.jx:199:5)
    at afterExtras (/usr/local/bin/node_modules/npm/lib/config/core.js.jx:190:5)
    at /usr/local/bin/node_modules/npm/node_modules/mkdirp/index.js.jx:47:26
    at Object.oncomplete (fs.js:82:15)
```

The files it is looking for are in the .jxp file

```
		"node_modules/npm/node_modules/npm-registry-client/lib/access.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/adduser.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/attempt.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/authify.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/deprecate.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/dist-tags/add.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/dist-tags/fetch.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/dist-tags/rm.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/dist-tags/set.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/dist-tags/update.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/fetch.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/get.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/initialize.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/logout.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/ping.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/publish.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/request.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/star.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/stars.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/tag.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/team.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/unpublish.js",
		"node_modules/npm/node_modules/npm-registry-client/lib/whoami.js",
```

Is there anything obviously wrong about what we are trying to do here?

**Comment by ktrzeciaknubisa** [comment]
> ENOENT, no such file or directory '/usr/local/bin/node_modules/npm/node_modules/npm-registry-client/lib'

Am I seeing right, that your kbox binary stays at `/usr/local/bin/` and you're running it from there?

**Comment by pirog** [comment]
yup `kbox` binary is in `/usr/local/bin` but its in $PATH so it can be run from anywhere. For the above i am not in `/usr/local/bin`.

**Comment by ktrzeciaknubisa** [comment]
Generally we were advising to run binary packages from their own location. See for example: https://github.com/jxcore/jxcore/issues/466. 

Could you please make a test and copy kbox to your app dir and run `./kbox` from there to see if this makes any difference?

**Comment by obastemur** [comment]
@ktrzeciaknubisa it should not be a problem using from path. At least its not a packaging issue. Or is it?

**Comment by pirog** [comment]
Yeah it works fine if i run it from the source directory but that makes sense since the source is all there and can be referenced relatively if needed. Effectively this also just means i still have dependencies :( 

Does this mean that any relative path ref ie `__dirname` is just a no-go with packaging?

**Comment by pirog** [comment]
@obastemur i'm kind of guessing this might be an issue with packaging. If you look at the offending code in the `npm-registry-client` module its

```json
fs.readdirSync(join(__dirname, 'lib')).forEach(function (f) {
    var entry = join(__dirname, 'lib', f)

    // lib/group-name/operation.js -> client.groupName.operation
    var stat = fs.statSync(entry)
    if (stat.isDirectory()) {
      var groupName = f.replace(/-([a-z])/, dashToCamel)
      fs.readdirSync(entry).forEach(function (f) {
        if (!f.match(/\.js$/)) return

        if (!client[groupName]) {
          // keep client.groupName.operation from stomping client.operation
          client[groupName] = Object.create(client)
        }
        var name = f.replace(/\.js$/, '').replace(/-([a-z])/, dashToCamel)
        client[groupName][name] = require(join(entry, f))
      })
      return
    }

    if (!f.match(/\.js$/)) return
    var name = f.replace(/\.js$/, '').replace(/-([a-z])/, dashToCamel)
    client[name] = require(entry)
  })```

Dynamic require based on `__dirname` seems like it could be an issue with packaging and then running from a different spot?

**Comment by ktrzeciaknubisa** [comment]
Relative paths inside a package are built against `process.cwd()`, however I don't remember any problems with `__dirname` (unless `join()` from snippet above does some tricks).

@pirog which npm version are you packing up? I don't find the code in npm 1.4.14 used by JXcore.

Also I don't get the `npm install` part. Do you mean `npm.commands.install()`, since you do it programmatically?

**Comment by pirog** [comment]
Yes! `npm.commands.install()` exactly

**Comment by ktrzeciaknubisa** [comment]
I was able to reproduce your error `ENOENT, no such file or directory` with npm 2.5.1, however I don't think we should worry much about that. That is because anyway maybe it's better if you use the exact npm used by JXcore. It contains multiple changes for JXcore compatibility (and this will be important if you want to run other things against jxcore). Also, I've just checked that it can be safely added into a native package and it works fine as well.

Just replace the *node_modules/npm* from your solution folder and put into there copied `~/.jx/npm`.  

Then your kbox file can still stay in PATH. I called it from another location and it worked.
